File: src/basango/services/http_client.py
import random
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from typing import Any, Optional, TypeAlias

import httpx

from basango.core.config import ClientConfig
from basango.services.user_agents import UserAgentProvider

HttpHeaders: TypeAlias = dict[str, str] | None
HttpParams: TypeAlias = dict[str, Any] | None
HttpData: TypeAlias = Any | None

TRANSIENT_STATUSES = (429, 500, 502, 503, 504)


@dataclass
class HttpClient:
    client_config: ClientConfig
    user_agent_provider: UserAgentProvider | None = None
    default_headers: HttpHeaders = None

    def _compute_backoff(self, attempt: int) -> float:
        base = min(
            self.client_config.backoff_initial
            * (self.client_config.backoff_multiplier**attempt),
            self.client_config.backoff_max,
        )
        jitter = random.uniform(0, base * 0.25)
        return base + jitter

    def _retry_delay(
        self, attempt: int, response: Optional[httpx.Response] = None
    ) -> float:
        delay = 0.0

        if response is not None and self.client_config.respect_retry_after:
            retry_after = (
                response.headers.get("Retry-After") if response.headers else None
            )
            if retry_after:
                try:
                    delay = max(0.0, float(int(retry_after)))
                except ValueError:
                    try:
                        dt = parsedate_to_datetime(retry_after)
                        if dt.tzinfo is None:
                            dt = dt.replace(tzinfo=timezone.utc)
                        now = datetime.now(timezone.utc)
                        delay = max(0.0, (dt - now).total_seconds())
                    except Exception:  # noqa: BLE001
                        pass

        if delay == 0.0:
            delay = self._compute_backoff(attempt)

        return delay

    def __post_init__(self) -> None:
        if self.user_agent_provider is not None:
            user_agent = self.user_agent_provider.get()
            self._user_agent = (
                user_agent if user_agent else self.client_config.user_agent
            )
        else:
            provider = UserAgentProvider(
                rotate=self.client_config.rotate,
                fallback=self.client_config.user_agent,
            )
            user_agent = provider.get()
            self._user_agent = (
                user_agent if user_agent else self.client_config.user_agent
            )

        headers = {"User-Agent": self._user_agent}

        if self.default_headers:
            headers.update(self.default_headers)

        self._client = httpx.Client(
            follow_redirects=self.client_config.follow_redirects,
            max_redirects=5,
            verify=self.client_config.verify_ssl,
            timeout=self.client_config.timeout,
            headers=headers,
        )

    # Context manager support -------------------------------------------------
    def __enter__(self) -> "HttpClient":  # noqa: D401
        return self

    def __exit__(self, exc_type, exc, tb) -> None:  # noqa: D401
        self.close()

    def close(self) -> None:
        try:
            self._client.close()
        except Exception:  # noqa: BLE001
            pass

    # Core request with retries ----------------------------------------------
    def _request(
        self,
        method: str,
        url: str,
        *,
        headers: HttpHeaders = None,
        params: HttpParams = None,
        data: Any | None = None,
        json: Any | None = None,
    ) -> httpx.Response:
        attempt = 0
        while True:
            try:
                response = self._client.request(
                    method, url, headers=headers, params=params, data=data, json=json
                )
                if (
                    response.status_code in TRANSIENT_STATUSES
                ) and attempt < self.client_config.max_retries:
                    time.sleep(self._retry_delay(attempt, response))
                    attempt += 1
                    continue
                response.raise_for_status()
                return response
            except httpx.HTTPStatusError as e:
                status = e.response.status_code if e.response else 0
                if (
                    status in TRANSIENT_STATUSES
                ) and attempt < self.client_config.max_retries:
                    time.sleep(self._retry_delay(attempt, e.response))
                    attempt += 1
                    continue
                raise
            except httpx.RequestError:
                if attempt < self.client_config.max_retries:
                    time.sleep(self._compute_backoff(attempt))
                    attempt += 1
                    continue
                raise

    # Public helpers ----------------------------------------------------------
    def get(self, url: str) -> httpx.Response:
        return self._request("GET", url)

    def post(
        self, url: str, data: HttpData = None, json: HttpData = None
    ) -> httpx.Response:
        return self._request("POST", url, data=data, json=json)

File: apps/crawler/src/http/http-client.ts
import { setTimeout as delay } from "node:timers/promises";

import { FetchClientConfig } from "#crawler/config";
import {
  DEFAULT_RETRY_AFTER_HEADER,
  DEFAULT_USER_AGENT,
  TRANSIENT_HTTP_STATUSES,
} from "#crawler/constants";
import { UserAgents } from "#crawler/http/user-agent";

export type HttpHeaders = Record<string, string>;
export type HttpParams = Record<string, string | number | boolean | null | undefined>;
export type HttpData = unknown;

export interface HttpClientOptions {
  userAgentProvider?: UserAgents;
  defaultHeaders?: HttpHeaders;
  fetchImpl?: typeof fetch;
  sleep?: (ms: number) => Promise<void>;
}

export interface HttpRequestOptions {
  headers?: HttpHeaders;
  params?: HttpParams;
  data?: HttpData;
  json?: HttpData;
  retryAfterHeader?: string;
}

export class HttpError extends Error {
  readonly status: number;
  readonly response: Response;

  constructor(message: string, response: Response) {
    super(message);
    this.status = response.status;
    this.response = response;
  }
}

/**
 * Default sleep function using setTimeout.
 * @param ms - Milliseconds to sleep
 */
const defaultSleep = (ms: number): Promise<void> => {
  return delay(ms).then(() => undefined);
};

/**
 * Builds a URL with query parameters.
 * @param url - The base URL
 * @param params - The query parameters to append
 */
const buildUrl = (url: string, params?: HttpParams): string => {
  if (!params || Object.keys(params).length === 0) {
    return url;
  }

  const target = new URL(url);
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null) continue;
    target.searchParams.set(key, String(value));
  }

  return target.toString();
};

/**
 * Computes the backoff time in milliseconds based on the configuration and attempt number.
 * @param config - Fetch client configuration
 * @param attempt - Current attempt number
 */
const computeBackoff = (config: FetchClientConfig, attempt: number): number => {
  const base = Math.min(
    config.backoffInitial * config.backoffMultiplier ** attempt,
    config.backoffMax,
  );
  const jitter = Math.random() * base * 0.25;
  return (base + jitter) * 1000;
};

const parseRetryAfter = (header: string): number => {
  const numeric = Number.parseInt(header, 10);
  if (!Number.isNaN(numeric)) {
    return Math.max(0, numeric * 1000);
  }

  const parsed = Date.parse(header);
  if (Number.isNaN(parsed)) {
    return 0;
  }

  const delta = parsed - Date.now();
  return delta > 0 ? delta : 0;
};

/**
 * Base HTTP client providing common functionality.
 *
 * @author Bernard Ngandu <bernard@devscast.tech>
 */
export class BaseHttpClient {
  protected readonly config: FetchClientConfig;
  protected readonly fetchImpl: typeof fetch;
  protected readonly sleep: (ms: number) => Promise<void>;
  protected readonly headers: HttpHeaders;

  constructor(config: FetchClientConfig, options: HttpClientOptions = {}) {
    this.config = config;
    const provider =
      options.userAgentProvider ??
      new UserAgents(config.rotate, config.userAgent ?? DEFAULT_USER_AGENT);
    const userAgent = provider.get() ?? config.userAgent ?? DEFAULT_USER_AGENT;

    const baseHeaders: HttpHeaders = { "User-Agent": userAgent };
    if (options.defaultHeaders) {
      Object.assign(baseHeaders, options.defaultHeaders);
    }

    this.headers = baseHeaders;
    this.fetchImpl = options.fetchImpl ?? fetch;
    this.sleep = options.sleep ?? defaultSleep;
  }

  protected buildHeaders(headers?: HttpHeaders): HeadersInit {
    return { ...this.headers, ...(headers ?? {}) };
  }

  protected async maybeDelay(
    attempt: number,
    response?: Response,
    retryAfterHeader: string = DEFAULT_RETRY_AFTER_HEADER,
  ): Promise<void> {
    let waitMs = 0;

    if (response) {
      const retryAfter = response.headers.get(retryAfterHeader);
      if (retryAfter && this.config.respectRetryAfter) {
        waitMs = parseRetryAfter(retryAfter);
      }
    }

    if (waitMs === 0) {
      waitMs = computeBackoff(this.config, attempt);
    }

    if (waitMs > 0) {
      await this.sleep(waitMs);
    }
  }
}

/**
 * Synchronous HTTP client with retry and timeout capabilities.
 *
 * @author Bernard Ngandu <bernard@devscast.tech>
 */
export class SyncHttpClient extends BaseHttpClient {
  async request(method: string, url: string, options: HttpRequestOptions = {}): Promise<Response> {
    const retryAfterHeader = options.retryAfterHeader ?? DEFAULT_RETRY_AFTER_HEADER;
    const target = buildUrl(url, options.params);

    const maxAttempts = this.config.maxRetries + 1;
    let attempt = 0;
    let lastError: unknown;

    while (attempt < maxAttempts) {
      const controller = new AbortController();
      let timeoutHandle: ReturnType<typeof setTimeout> | undefined;
      try {
        timeoutHandle = setTimeout(() => controller.abort(), this.config.timeout * 1000);

        const headers = this.buildHeaders(options.headers);
        const init: RequestInit = {
          body: options.data as BodyInit | undefined,
          headers,
          method,
          redirect: this.config.followRedirects ? "follow" : "manual",
          signal: controller.signal,
        };

        if (options.json !== undefined) {
          init.body = JSON.stringify(options.json);
          (init.headers as Record<string, string>)["Content-Type"] ??= "application/json";
        }

        const response = await this.fetchImpl(target, init);

        if (
          TRANSIENT_HTTP_STATUSES.includes(response.status as number) &&
          attempt < this.config.maxRetries
        ) {
          await this.maybeDelay(attempt, response, retryAfterHeader);
          attempt += 1;
          continue;
        }

        if (!response.ok) {
          throw new HttpError(`HTTP ${response.status} ${response.statusText}`, response);
        }

        return response;
      } catch (error) {
        if (error instanceof HttpError) {
          lastError = error;
          throw error;
        }

        if (error instanceof DOMException && error.name === "AbortError") {
          lastError = error;
          if (attempt >= this.config.maxRetries) {
            throw error;
          }
        } else {
          lastError = error;
          if (attempt >= this.config.maxRetries) {
            throw error;
          }
        }

        await this.maybeDelay(attempt);
        attempt += 1;
      } finally {
        if (timeoutHandle) {
          clearTimeout(timeoutHandle);
        }
      }
    }

    throw lastError instanceof Error ? lastError : new Error("HTTP request failed after retries");
  }

  get(url: string, options?: Omit<HttpRequestOptions, "data" | "json">): Promise<Response> {
    return this.request("GET", url, options);
  }

  post(url: string, options: HttpRequestOptions = {}): Promise<Response> {
    return this.request("POST", url, options);
  }
}

export type HttpClient = SyncHttpClient;
